Do not overwrite TMPDIR

aaronjensen opened the pull request:
Fixes #13325 

Maybe a check for 'nobody' would be more appropriate, I don't know, but this existing behavior is surprising and causes mysterious/confusing problems.

othiym23 commented:
As the comment indicates, just straight-up deleting those lines is going to cause problems for users where bash can't write to `$TMPDIR` as `nobody`. Having some kind of tests as to whether the CLI can write to the current directory and / or `$TMPDIR` and choosing one or the other (and maybe bailing out if there's _nowwhere_ we can write to) seems much less likely to cause unexpected regressions. Also, this needs a test to be merged. Thanks for the work you've done thus far!

othiym23 commented:
See also d7dcacea6f17437fd88321987ca6aa8fe3c017d8, when this behavior was added: `/bin/sh` on Solaris / SmartOS systems, which is "pure" Bourne shell, freaks out when it can't write to TMPDIR. Cross-platform compatibility! 💫

aaronjensen committed:
Use TMPDIR if it is writable

Only fall back to working directory if TMPDIR is not writable, error if
neither are writable.

aaronjensen commented:
@othiym23 Thank you for the feedback. Please see the updated commit and let me know if there's anything I can do to improve it.

aaronjensen commented on lib/utils/lifecycle.js:
```diff
@@ -75,18 +76,38 @@ function lifecycle (pkg, stage, wd, unsafe, failOk, cb) {
+    return true
+  } catch (e) {
+    return false
+  }
```
I'm not a fan of using exceptions in this way, but it felt like an awkward place to have to use callbacks. If you'd prefer that let me know and I can switch it out.

iarna replied:
In a test, this is fine (I do it myself).

aaronjensen commented on test/tap/lifecycle.js:
```diff
@@ -42,3 +42,30 @@ test('lifecycle : rejects wd for ', function (t) {
   })
 })
+
+test('lifecycle : accepts TMPDIR that is writable', function (t) {
```
These tests assume TMPDIR is set properly and writeable, hopefully that's not a problem anywhere.

aaronjensen commented on test/tap/lifecycle.js:
```diff
@@ -42,3 +42,30 @@ test('lifecycle : rejects wd for ', function (t) {
+  var wd = '/probably/not/writable/either'
+
+  lifecycle._getTemporaryDirectory(wd, {TMPDIR: '/probably/not/writable'}, function (er, tmpdir) {
+    t.type(er, Error)
```
I'm new to tap and `t.error` didn't seem to do what I expected. Is this the proper way to test this?

iarna replied:
Yeah `t.ifError` (and `t.error`) go against the `tap` grain and assert there WASN'T an error.

I would just do: `t.ok(er, 'assertion description')` to assert that an error _was_ produced.

aaronjensen commented:
Thanks @iarna. I updated the test to use `t.ok`. Is there anything else you'd like me to update on this or is it good to go?

iarna reviewed (approved):
LGTM

iarna commented:
This does not pass on 0.10 because it uses `fs.accessSync` which is not available on 0.10. We've gotta support 0.10 a little longer so I'm going to defer merging it.

iarna reviewed (changes requested):
There's an issue with this: It checks to see if `TMPDIR` is writable _by the current user_. But the error condition we're trying to guard against is `TMPDIR` not being writable by the uid & gid passed in.

I'm not sure there's a meaningful way to check for that, short of spawning a sub process to check for us.

iarna commented:
As I ultimately found a problem with this that I don't know how to resolve, and there's been no reply in over six months, I'm going to close this. Please feel free to open a new PR if you'd like to continue this discussion. Thanks!
